pkg/gui/controllers/files_controller.go

```diff
@@ -302,49 +302,61 @@ func (self *FilesController) stageAll() error {
 	return self.contexts.Files.HandleFocus()
 }
 
-func (self *FilesController) ignore(node *filetree.FileNode) error {
-	if node.GetPath() == ".gitignore" {
-		return self.c.ErrorMsg("Cannot ignore .gitignore")
-	}
+func (self *FilesController) unstageFiles(node *filetree.FileNode) error {
 
-	unstageFiles := func() error {
-		return node.ForEachFile(func(file *models.File) error {
-			if file.HasStagedChanges {
-				if err := self.git.WorkingTree.UnStageFile(file.Names(), file.Tracked); err != nil {
-					return err
-				}
+	return node.ForEachFile(func(file *models.File) error {
+		if file.HasStagedChanges {
+			if err := self.git.WorkingTree.UnStageFile(file.Names(), file.Tracked); err != nil {
+				return err
 			}
+		}
 
-			return nil
-		})
-	}
+		return nil
+	})
+}
+
+func (self *FilesController) checkTracking(node *filetree.FileNode, trText string, trPrompt string, trAction string, f func(string) error) error {
 
 	if node.GetIsTracked() {
 		return self.c.Confirm(types.ConfirmOpts{
-			Title:  self.c.Tr.IgnoreTracked,
-			Prompt: self.c.Tr.IgnoreTrackedPrompt,
+			Title:  trText,
+			Prompt: trPrompt,
 			HandleConfirm: func() error {
-				self.c.LogAction(self.c.Tr.Actions.IgnoreFile)
+				self.c.LogAction(trAction)
 				// not 100% sure if this is necessary but I'll assume it is
-				if err := unstageFiles(); err != nil {
+				if err := self.unstageFiles(node); err != nil {
 					return err
 				}
 
 				if err := self.git.WorkingTree.RemoveTrackedFiles(node.GetPath()); err != nil {
 					return err
 				}
 
-				if err := self.git.WorkingTree.Ignore(node.GetPath()); err != nil {
+				if err := f(node.GetPath()); err != nil {
 					return err
 				}
+
 				return self.c.Refresh(types.RefreshOptions{Scope: []types.RefreshableView{types.FILES}})
 			},
 		})
 	}
+	return nil
+}
+
+func (self *FilesController) ignore(node *filetree.FileNode) error {
+	if node.GetPath() == ".gitignore" {
+		return self.c.ErrorMsg("Cannot ignore .gitignore")
+	}
 
-	self.c.LogAction(self.c.Tr.Actions.IgnoreFile)
+	trackingErr := self.checkTracking(node, self.c.Tr.IgnoreTracked, self.c.Tr.IgnoreTrackedPrompt, self.c.Tr.Actions.IgnoreExcludeFile, self.git.WorkingTree.Ignore)
```

Comment:
checkTracking should return both an error and a bool, so that if the bool is true we early-exit from this function.

Reply:
Sure :) `checkTracking` is unlikely to return an error. If `node.GetIsTracked()` is true, then we'll return `self.c.Confirm()` but that method is also unlikely to return an error. `self.c.Confirm()` also doesn't block so it'll return basically immediately.

This means that when we call `checkTracking` from within `ignore`, regardless of whether it brings up a confirmation popup, we'll continue on then unstage the files and ignore the given path, as if the confirmation popup had not been opened.

So I'm suggesting we add a bool alongside the error that tells us whether the popup was opened, for the sake of exiting early if it was. So we can first check the error and bubble it up if it's present, and we can secondly check the bool to see if we should early exit. Although we could encode the need for an early exit in the error itself, rather than using a separate bool, that would be an abuse of the error type because errors should only capture 'unexpected' events.

Lemme know if that makes sense

Reply:
Yeah that makes sense. I'll update this code with that when I get back to my laptop :)